### PR TITLE
Update 0.2.1

### PR DESCRIPTION
**Pull request by Shawdooow**
-Featuring SeekingBullets and the Helper object
-Enemies buffed to resist extra bullets flying around now

### CODE REVIEW

**Comment by Jorolf** [comment]
🤔 you should probably keep osu merges and vitaru merges seperate
to make it easier to review

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            return BulletVelocity;
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+            Rotation = Rotation + 1f;
+
+            //IdleTimer
+            if(startTime + 300 <= Time.Current)
```
Comment (Jorolf) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: what does this?

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            bulletRing.ScaleTo(new Vector2(1), 150, EasingTypes.OutCubic);
+            enemyPos = MathHelper.DegreesToRadians(StartAngle - 90);
+            GetBulletVelocity();
+        }
+
+        private void nearestEnemy()
+        {
+            if (VitaruPlayfield.vitaruPlayfield != null && VitaruBeatmapConverter.EnemyList != null)
+            {
+                foreach (Drawable draw in VitaruBeatmapConverter.EnemyList)
```
Comment (Jorolf) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: Should the enemy list only contain enemies?

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                {
+                    if (draw is DrawableVitaruEnemy)
+                    {
+                        drawableVitaruEnemy = draw as DrawableVitaruEnemy;
```
Comment (Jorolf) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: Also for stuff like this:
try to minimize the use of is and as
do as and then null check

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                    },
+                    EdgeEffect = new EdgeEffect
+                    {
+                        Type = EdgeEffectType.Shadow,
+                        Radius = BulletWidth / 2,
+                        Colour = BulletColor.Opacity(0.25f),
+                    },
+                },
+            };
+            bulletRing.FadeIn(150, EasingTypes.OutCubic);
```
Comment (Jorolf) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: 🤔 doesn't FadeInFromZero exist? or did they remove that

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs (hunk 1 above)
Comment (Shawdooow) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: makes them fly straight for a moment when spawning

Comment (Jorolf) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: try Scheduler and a boolean

Comment (Jorolf) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: so you don't check every update

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs (hunk 4 above)
Comment (Shawdooow) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: Does it make a difference if it starts at `Alpha = 0` regardless?

Comment (Jorolf) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: you dont have to write Alpha = 0f ?

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs (hunk 3 above)
Comment (Shawdooow) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: This is the same way for checking as for `HitDetect` in `DrawableCharacter`. what would the difference be?

Comment (Jorolf) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: one less cast

File: osu.Game.Rulesets.Vitaru/Objects/Helper.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Colour;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Rulesets.Vitaru.Objects.Projectiles;
+using osu.Game.Rulesets.Vitaru.UI;
+
+namespace osu.Game.Rulesets.Vitaru.Objects.Drawables
+{
+    public class Helper : Container
```
Comment (ColdVolcano) [inline osu.Game.Rulesets.Vitaru/Objects/Helper.cs]: At least xmldoc what even a helper is

File: osu.Game.Rulesets.Vitaru/Objects/Drawables/DrawableVitaruPlayer.cs [hunk 6]
```diff
@@ -58,13 +58,22 @@ protected override void CheckJudgement(bool userTriggered)
...
         protected override void Update()
         {
             base.Update();
 
             HitDetect();
 
+            if (helper1.Alpha == 0)
+            {
+                helper1.Position = new Vector2(20, -30);
+                helper2.Position = new Vector2(-20, -30);
```
Comment (ColdVolcano) [inline osu.Game.Rulesets.Vitaru/Objects/Drawables/DrawableVitaruPlayer.cs]: why do we need to set a constant position on every update?

File: osu.Game.Rulesets.Vitaru/Objects/Helper.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                    },
+                },
+                helperContainer = new CircularContainer
+                {
+                        Origin = Anchor.Centre,
+                        Anchor = Anchor.Centre,
+                        RelativeSizeAxes = Axes.Both,
+                        Scale = new Vector2(HelperWidth * 2),
+                        Depth = 2,
+                        Masking = true,
```
Comment (ColdVolcano) [inline osu.Game.Rulesets.Vitaru/Objects/Helper.cs]: da spacing is real

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            bulletRing.FadeInFromZero(150, EasingTypes.OutCubic);
+            bulletRing.ScaleTo(new Vector2(1), 150, EasingTypes.OutCubic);
+            enemyPos = MathHelper.DegreesToRadians(StartAngle - 90);
+            GetBulletVelocity();
+        }
+
+        private void nearestEnemy()
+        {
+            if (VitaruBeatmapConverter.EnemyList != null)
+            {
```
Comment (ColdVolcano) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: no need for this much indentation, just return if it is null

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+                foreach (DrawableVitaruEnemy enemy in VitaruBeatmapConverter.EnemyList.OfType<DrawableVitaruEnemy>())
+                {
+                    if(enemy.Alpha > 0)
+                    {
```
Comment (ColdVolcano) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: if (enemy.Alpha<=0)
continue;

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                        float minDist = 9999;
+                        float dist = Vector2.Distance(enemy.Position, Position);
+                        if (dist < minDist)
```
Comment (ColdVolcano) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: if (dist>=mindist)
continue;

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        {
+            base.Update();
+            Rotation = Rotation + 1f;
+
+            //IdleTimer
+            if(startTime + 300 <= Time.Current)
+            {
+                nearestEnemy();
+                if(NearestEnemy != null)
+                    if (NearestEnemy.CharacterHealth > 0)
```
Comment (ColdVolcano) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: what about these two if statements in one?

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs (hunk 9 above)
Comment (Shawdooow) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: This seems like a preference thing more than a problem?

File: osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs (hunk 10 above)
Comment (Shawdooow) [inline osu.Game.Rulesets.Vitaru/Objects/Projectiles/SeekingBullet.cs]: Same as above?

**Pull request merged**